### PR TITLE
Bug fix/fix remote executor races

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Fix some races in the RemoteExecutor

- [X] Bug-Fix for *devel-branch* (no need for backports)

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6661/

File: arangod/Aql/RemoteExecutor.cpp [hunk 1]
```diff
@@ -442,46 +446,64 @@ Result ExecutionBlockImpl<RemoteExecutor>::sendAsyncRequest(fuerte::RestVerb typ
...
                                              std::unique_ptr<fuerte::Response> res) {
-                      _query.sharedState()->execute([&] {  // notifies outside
-                        std::lock_guard<std::mutex> guard(_communicationMutex);
-                        if (_lastTicket == ticket) {
-                          if (err != fuerte::Error::NoError || res->statusCode() >= 400) {
-                            _lastError = handleErrorResponse(spec, err, res.get());
-                          } else {
-                            _lastResponse = std::move(res);
-                          }
+                      std::lock_guard<std::mutex> guard(_communicationMutex);
```
Comment (graetzer) [inline arangod/Aql/RemoteExecutor.cpp]: actually I do not get the need for the extra mutex here, an atomic _lastTicket should be sufficient so synchronize access to _lastError and _lastResponse ?

File: arangod/Aql/RemoteExecutor.cpp [hunk 2]
```diff
@@ -351,17 +349,22 @@ std::pair<ExecutionState, Result> ExecutionBlockImpl<RemoteExecutor>::shutdown(i
 
     return {ExecutionState::DONE, TRI_ERROR_INTERNAL};
   }
-#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
-  TRI_ASSERT(_didSendShutdownRequest == false);
+
+  // Already sent a shutdown request, but haven't got an answer yet.
+  if (_didSendShutdownRequest) {
+    return {ExecutionState::WAITING, TRI_ERROR_NO_ERROR};
```
Comment (mchacki) [inline arangod/Aql/RemoteExecutor.cpp]: Do we want to add. traceShutdown here as well?

File: arangod/Aql/RemoteExecutor.cpp (hunk 1 above)
Comment (goedderz) [reply arangod/Aql/RemoteExecutor.cpp]: The problem is that we do not only have to synchronize a request with its answer, but also protect an answer with another request; e.g., RemoteExecutor sends a getSome/skipSome request, then a shutdown happens (e.g. due to a timeout) at the same time when the answer arrives. Then, this code in `ExecutionBlockImpl<RemoteExecutor>::shutdown`
```
  if (!_hasTriggeredShutdown) {
    std::lock_guard<std::mutex> guard(_communicationMutex);
    std::ignore = generateNewTicket();
    _hasTriggeredShutdown = true;
  }
```
which resets `_lastTicket`, `_lastError` and `_lastResponse`, races with the lambda that's called with the answer:
```
                    [=, ref(std::move(ref))](fuerte::Error err,
                                             std::unique_ptr<fuerte::Request>,
                                             std::unique_ptr<fuerte::Response> res) {
                      std::lock_guard<std::mutex> guard(_communicationMutex);

                      if (_lastTicket == ticket) {
                        if (err != fuerte::Error::NoError || res->statusCode() >= 400) {
                          _lastError = handleErrorResponse(spec, err, res.get());
                        } else {
                          _lastResponse = std::move(res);
                        }
                        _query.sharedState()->execute();
                      }
                    }
```
which reads `_lastTicket` and then sets `_lastError` or `_lastResponse`. I do not see how that should be correctly synchronized with just an atomic `_lastTicket`.

File: arangod/Aql/RemoteExecutor.cpp (hunk 2 above)
Comment (goedderz) [reply arangod/Aql/RemoteExecutor.cpp]: From the code I had locally, I just added the tracing of actual remote requests, not the tracing of `::shutdown()` calls - I thought that'd be enough in most of the cases where shutdown is of interest at all, and doesn't clutter the log as much.

**Pull request merged**